### DISCUSSION TITLE
Multithreading and Cancellation

### DISCUSSION THREAD

**Discussion by gregjohnson2017** [andreiavrammsd/cpp-channel#32] [Q&A]
I'm trying to use the channels that this project offers to communicate between threads in a project that I'm working on. However, I'm wondering how to use the std::stop_token cancellation feature of thread/jthread when my runner thread is blocked trying to read from a channel. 

I ran `examples/multithreading.cpp` to see an example of threads using channels to communicate and then being gracefully terminated, but unfortunately this example appears to never terminate. The example code calls join on every reader and the writer, however the main thread will just block on the first join call, since none of the writers/readers are ever cancelled. As a side note, I believe there is also a race condition in how std::cout is being written to concurrently

**Comment by andreiavrammsd** [comment]
The channel blocks and waits for data to be written. It's by design.

What you can do is [close the channel](https://github.com/andreiavrammsd/cpp-channel/blob/master/examples/close.cpp). Then you cannot write to it any more and it will unblock. Though the `close` feature might have issues in some cases.

The [multithreading example](https://github.com/andreiavrammsd/cpp-channel/blob/master/examples/multithreading.cpp) never terminates because that's the use case it presents: a long-running program that continuously passes data into the channel and reads it from. Indeed, the `Wait for all threads to finish` comment in that file is not correct and I'll fix it.
